### PR TITLE
Add support for TLS config command-line options

### LINKED ISSUE #1716: Add support for -H and TLS flags
The `docker-compose` command should support these options from the `docker` command:

```
  -H, --host=[]                      Daemon socket(s) to connect to 
  --tls=false                        Use TLS; implied by --tlsverify
  --tlscacert=~/.docker/ca.pem       Trust certs signed only by this CA
  --tlscert=~/.docker/cert.pem       Path to TLS certificate file
  --tlskey=~/.docker/key.pem         Path to TLS key file
  --tlsverify=false                  Use TLS and verify the remote
```

This will allow you to do `docker-compose $(docker-machine config dev) ...` etc.

### PR DESCRIPTION
**Pull request by shin-**
Fixes #1716

### CODE REVIEW

File: compose/cli/main.py [hunk 1]
```diff
@@ -147,6 +147,13 @@ class TopLevelCommand(object):
       -p, --project-name NAME   Specify an alternate project name (default: directory name)
       --verbose                 Show more output
       -v, --version             Print version and exit
+      -H, --host HOST           Daemon socket to connect to
+
+      --tls                     Use TLS; implied by --tlsverify
+      --tlsacert                Trust certs signed only by this CA
```
Comment (tombee) [inline compose/cli/main.py]: Hey, should this be `--tlscacert`? :)

**Comment by aanand** [comment]
Nice. Rather than create an intermediate `TLSArgs` object, how about just having a method `tls_config_from_options` which takes the options dict and returns a `TLSConfig` object? Then `docker_client()` can just take a `tls_config` kwarg.

**Comment by tombee** [comment]
I've been trying this out with just TLS verification using the CA cert, but I can't seem to get it to work (same thing happens with and without `-H`).

```
$ docker-compose -H 192.168.1.16:443 --tlsverify --tlscacert ca.pem up
$ docker-compose --tlsverify --tlscacert ca.pem up

Define and run multi-container applications with Docker.

Usage:
  docker-compose [-f=<arg>...] [options] [COMMAND] [ARGS...]
  docker-compose -h|--help

Options:
  -f, --file FILE           Specify an alternate compose file (default: docker-compose.yml)
  -p, --project-name NAME   Specify an alternate project name (default: directory name)
  --verbose                 Show more output
  -v, --version             Print version and exit
  -H, --host HOST           Daemon socket to connect to

  --tls                     Use TLS; implied by --tlsverify
  --tlsacert                Trust certs signed only by this CA
  --tlscert                 Path to TLS certificate file
  --tlskey                  Path to TLS key file
  --tlsverify               Use TLS and verify the remote
```

**Comment by aanand** [comment]
@tombee I think that's because of the typo you found - try passing `--tlsacert`.

I wish we could show a better error message when an arg isn't recognised.

**Comment by tombee** [comment]
@aanand Tried that, I think the option is actually `--tlscacert`: https://github.com/docker/compose/pull/3139/files#diff-25307ba665155091ce1d48c75283dd6aR24 :)

Think it's just our help message that had the typo?

**Comment by aanand** [comment]
@tombee We use docopt, so our help message is what the command line arg parser is generated from!

**Comment by shin-** [comment]
Sorry, it's still wip. I'll push an update soon!

**Comment by shin-** [comment]
Should look better now. @tombee if you wanna give it another shot!

File: compose/cli/docker_client.py [hunk 2]
```diff
@@ -14,7 +15,27 @@
...
-def docker_client(version=None):
+def tls_config_from_options(options):
+    tls = options.get('--tls', False)
+    ca_cert = options.get('--tlscacert')
+    cert = options.get('--tlscert')
+    key = options.get('--tlskey')
+    verify = options.get('--tlsverify')
+
+    if tls is True:
+        return True
```
Comment (dnephin) [inline compose/cli/docker_client.py]: Setting `--tls` makes it ignore all the other tls flags? is that how the docker-cli works?

Comment (shin-) [inline compose/cli/docker_client.py]: I thought so, but checking again, it seems another valid option is `docker --tls --tlscert ... --tlskey ...`  which means "Authenticate with client certificate, do not authenticate server based on given CA" (what we would do here if `--tls` was omitted). I'll update to match.

**Comment by tombee** [comment]
If I prefix with `tcp://`, then TLS doesn't seem to get enabled:

```
$ docker-compose -H tcp://192.168.1.16:443 --tlsverify --tlscacert ca.pem up
ERROR: Couldn't connect to Docker daemon at http://192.168.1.16:443 - is it running?
```

**Comment by tombee** [comment]
Thanks for this @shin-!  Making a bit of progress, but still unable to use our CA cert for verification.

```
$ docker-compose -H 192.168.1.16:443 --tlsverify --tlscacert ca.pem up
ERROR: SSL error: hostname '192.168.1.16' doesn't match 'debian'
```

I've verified that I can use the same configuration with the engine CLI:

```
$ docker -H 192.168.1.16:443 --tlsverify --tlscacert ca.pem info
Containers: 23
Images: 24
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 1
 debian: 192.168.1.16:12376
...
```

Also seeing that I get a slightly different error if `--tls` is supplied even though our docs say `--tls=false - Use TLS; implied by --tlsverify`, so I was expecting the behaviour to be the same :)

```
$ docker-compose -H 192.168.1.16:443 --tls --tlsverify --tlscacert ca.pem up
ERROR: SSL error: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)
```

**Comment by shin-** [comment]
- The `tcp://` address not being translated to an `https` endpoint is a docker-py bug.
- This `ERROR: SSL error: hostname '192.168.1.16' doesn't match 'debian'` is a case of `assert_hostname` being enabled by default. Don't know if we should disable it by default (docker engine seems to ignore it?) or add an option to do so.
- The issue with the `--tls` flag is linked to what Daniel mentions here https://github.com/docker/compose/pull/3139#discussion_r56447620 .

Thanks for your help with testing!

**Comment by aanand** [comment]
We should definitely not check the hostname by default, as that'll break all Docker Machine local VMs. We could put an option in to enable hostname checking, but if the CLI doesn't do it, I think it's pretty low-priority.

**Comment by shin-** [comment]
I left the option to disable the hostname check entirely, and modified the code a little bit to pass the user-specified hostname as the `assert_hostname`. It's not well documented but it does seem to be a feature of urllib3 / requests. This should work as expected, but if someone could test it with docker-machine that would be super helpful.

**Comment by dnephin** [comment]
Looks like there is a python2.7 issue

**Comment by dnephin** [comment]
LGTM, the one test flake is fixed in master now, if you want to rebase

File: tests/unit/cli_test.py [hunk 3]
```diff
@@ -20,6 +21,85 @@
 from compose.project import Project
 
 
+class TLSConfigTestCase(unittest.TestCase):
```
Comment (dnephin) [inline tests/unit/cli_test.py]: hmm, I think this should  be in `tests/unit/cli/docker_client_test.py`

**Comment by dnephin** [comment]
LGTM

**Comment by tombee** [comment]
@shin- Awesome work on this btw!

**Pull request merged**